add modularization overview (#711)
kyma-project/community: concepts/modularization/modularization_overview.md

---
title: Modularization in Kyma
---

With Kyma’s modular approach, you can install exactly the Kyma components you want, making the installation very lightweight and adjusted to your business needs. Kyma provides three components to support the modular approach: The Module Manager, the Lifecycle Manager, and the Runtime Watcher.

## Kyma’s Modular Approach

![](assets/Modular_Approach.svg)

## 1. Kyma Module

A Kyma module contains all information required to install and run the associated components in a Kyma runtime:

* A renderable Kubernetes manifest that includes the Kubernetes resources of the module
* A default configuration for module chart resource installation
* Optionally, further information

All these assets are bundled into a single container image using the OCI image specification.

## 2. Module Manager

> The Module Manager installs and manages the Kyma components in a local or remote setup. For example, it remotely manages and synchronizes the installation in the SAP Kyma Runtime. The Module Manager retrieves the container image and processes the Kubernetes manifest by calling a rendering framework like Helm or Kustomize. Finally, it deploys the manifest in Kubernetes and monitors the outcome of the deployment.

Reviewer: ```suggestion
The Module Manager installs, uninstalls and manages the Kyma modules in a local (single cluster mode) or remote setup. E.g., it remotely manages and synchronizes Kyma module manifest resources in the SAP Kyma Runtime. The Module Manager retrieves the image layers from the specified image registry and processes the manifest resources by calling a rendering framework like Helm or Kustomize. Finally, it deploys this resource along with a custom resource to track state changes. It ensures consistency both via a time-based reconciliation and watching on state changes of the deployed custom resource in Kyma Runtime.
```

Author: Just minor language editing to @adityabhatia's suggestion:

```suggestion
The Module Manager installs, uninstalls, and manages the Kyma modules in a local (single cluster mode) or remote setup. For example, it remotely manages and synchronizes Kyma module manifest resources in the SAP Kyma Runtime. The Module Manager retrieves the image layers from the specified image registry and processes the manifest resources by calling a rendering framework like Helm or Kustomize. Finally, it deploys this resource along with a custom resource to track state changes. It ensures consistency both with a time-based reconciliation and by watching state changes of the deployed custom resource in Kyma runtime.
```

## 3. Runtime Watcher

> The Runtime Watcher monitors the relevant Kyma resources of the user’s Kyma runtime for any changes, such as:

Reviewer: ```suggestion
The Runtime Watcher monitors the relevant Kyma resources of the user’s Kyma runtime for configured changes, specified by the operator in the control plane, such as:
```

* Expected changes, like a Kyma user providing a configuration update, which must lead to an update of the particular Kyma runtime.

> * Unexpected changes, like an accidental deletion or modification, which must be reverted to recover the Kyma runtime to a healthy state.

Reviewer: ```suggestion
* Unexpected changes, like an accidental creation, deletion or modification, which must be reverted to recover the Kyma runtime to a healthy state.
```

If needed, the Runtime Watcher triggers a reconciliation of the user’s Kyma runtime by propagating the desired changes to the Module Manager and the Lifecycle Manager.

## 4. Lifecycle Manager

> The Lifecycle Manager is responsible for managing the lifecycle of Kyma instances, like installing or deleting a Kyma instance from a cluster. It checks whether a CR is provided for every module, and deletes obsoletes module if they’re no longer required

Reviewer: ```suggestion
The Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. This is done by generating required Kyma module custom resources, which bootstrap Kyma module processing. To add, it reconciles a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster. 
```

Author: Slight editing for @adityabhatia's suggestion

```suggestion
The Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. This is done by generating the required Kyma module custom resources, which bootstrap Kyma module processing. Furthermore, the Lifecycle Manager reconciles a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster. 
```

Author: @adityabhatia, what **_are_** module operators? It's the first time we mention the term here. We need to explain it in this overview doc.

Reviewer: Module operators are basically component operators, which will be installed by other Kyma component teams or third party component owners. These operators are responsible for managing the module (component) specific resources on SKR

Author: Because you wrote "basically" - are there other kinds of component operators? Or is "module operator" synonym to "component operator"?

Reviewer: Yes, we decided to call it module instead of a component, that is the reason we call it a module-operator

Reviewer 2: Module Operators are Controllers that steer the lifecycle of a single Module. A Module can consist of one or many compoents.

Example:
Telemetry Operator (Operator) is in charge of reconciling LogPipeline and other CRs. It controls the Kyma Components "Logging", "Monitoring" and "Tracing".

Author: Thank you! Where would module operators show up in the diagram? Where do they belong in the text?

Reviewer 2: They are abstracted under "Kyma Component". Technically it can also be a pure helm chart without an operator, but this is not recommended by the architecture design right now.

---
title: Modularization in Kyma
---

> With Kyma’s modular approach, you can install exactly the Kyma components you want, making the installation very lightweight and adjusted to your business needs. Kyma provides three components to support the modular approach: The Module Manager, the Lifecycle Manager, and the Runtime Watcher.

Reviewer 2: ```suggestion
With Kyma’s modular approach, you can install exactly the Kyma components you want, making the installation very lightweight and adjusted to your business needs. Kyma uses three system components to support the modular approach: The Module Manager, the Lifecycle Manager, and the Runtime Watcher.
```

## Kyma’s Modular Approach

> ![](assets/modular_approach.svg)

Reviewer 2: I think it is misleading to have the `Kyma CR` outside of both Control Plane and Kyma Cluster. The reason for this is that the `Kyma Instance` is in actuality the Kyma CR (they are the same). And there is an equivalent in Control Plane as well as in the Runtime.

On the other side, the `Kyma Module` has no `Config Chart`, instead the `Manifest CR` is containing image layers of which one of them is a Helm chart where another one is a configuration file. 

The Manifest is inside of the Control-Plane, and not outside like it is depicted in the graphic

Author: I tried implementing your suggestions with https://github.com/kyma-project/community/pull/711/commits/c8ab8c3b66743e8a681be4ee07c330beec60f900 but I'm pretty sure that's even less accurate. 
Could one of you try to adjust the image like it should be?

Author: 😮 nice!

## 1. Kyma Module

A Kyma module contains all information required to install and run the associated components in a Kyma runtime:

> * A renderable Kubernetes manifest that includes the Kubernetes resources of the module

Reviewer 2: ```suggestion
* One or many renderable Kubernetes manifests of the module, based on Kustomize and / or Helm
```

Author: ```suggestion
* One or more renderable Kubernetes manifests of the module, based on Kustomize and/or Helm
```

* A default configuration for module chart resource installation

> * Optionally, further information

Reviewer 2: ```suggestion
* Optionally, further layers that can enhance or be required by the operator of a module
```

Author: ```suggestion
* Optionally, further layers that can enhance, or be required by, the operator of a module
```

All these assets are bundled into a single container image using the OCI image specification.

> **TIP:** Learn how to build a module(ADD_LINK).

## 2. Module Manager

> The Module Manager installs, uninstalls, and manages the Kyma modules in a local (single cluster mode) or remote setup. For example, it remotely manages and synchronizes Kyma module manifest resources in the SAP Kyma Runtime.

Reviewer 2: ```suggestion
The Module Manager installs, uninstalls, and manages the Kyma modules in a local (single cluster mode) or remote setup (in which one cluster acts as control-plane). For example, it remotely manages and synchronizes Kyma module resources in the SAP Kyma Runtime while the Manifest is located in the Control Plane.
```

Author: We should standardize the spelling of "control plane" - capitalized or not? Is it a Kubernetes resource?

```suggestion
The Module Manager installs, uninstalls, and manages the Kyma modules in a local (single cluster mode) or remote setup (in which one cluster acts as control plane). For example, it remotely manages and synchronizes Kyma module resources in the SAP Kyma Runtime, while the manifest is located in the control plane.
```

The Module Manager retrieves the image layers from the specified image registry and processes the manifest resources by calling a rendering framework like Helm or Kustomize.

Finally, it deploys this resource along with a custom resource to track state changes. It ensures consistency both with a time-based reconciliation and by watching state changes of the deployed custom resource in Kyma runtime.

## 3. Runtime Watcher

> The Runtime Watcher monitors the relevant Kyma resources of the user’s Kyma runtime for configured changes, specified by the operator in the control plane, such as:

Reviewer 2: ```suggestion
The Runtime Watcher monitors the relevant Module resources of the user’s Kyma runtime for configured changes, specified by the operator in the control plane, such as:
```

Author: ```suggestion
The Runtime Watcher monitors the relevant module resources of the user’s Kyma runtime for configured changes, specified by the operator in the control plane, such as:
```

> * Expected changes, like a Kyma user providing a configuration update, which must lead to an update of the particular Kyma runtime.

Reviewer 2: ```suggestion
* Expected changes, like a Kyma user providing a configuration update to the module, which must lead to an update of related resources in the Kyma runtime.
```

* Unexpected changes, like an accidental creation, deletion, or modification, which must be reverted to recover the Kyma runtime to a healthy state.

> If needed, the Runtime Watcher triggers a reconciliation of the user’s Kyma runtime by propagating the desired changes to the Module Manager and the Lifecycle Manager.

Reviewer 2: ```suggestion
For all detected changes (see above), the Runtime Watcher triggers a reconciliation of the user’s Kyma runtime by propagating the desired changes to the Module Manager and the Lifecycle Manager.
```

Author: ```suggestion
For all such detected changes, the Runtime Watcher triggers a reconciliation of the user’s Kyma runtime by propagating the desired changes to the Module Manager and the Lifecycle Manager.
```

> 

Reviewer 2: ```suggestion
It should be noted that the Runtime Watcher is only a necessary component when running with a Control Plane managing one or many other Runtimes. For local (single-cluster) setups, the Runtime Watcher is not needed.
```

Author: ```suggestion
> **NOTE:** The Runtime Watcher is a necessary component only when running with a control plane that manages one or more other runtimes. For local (single-cluster) setups, the Runtime Watcher is not needed.
```

## 4. Lifecycle Manager

> The Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. This is done by generating the required Kyma module custom resources, which bootstrap Kyma module processing. Furthermore, the Lifecycle Manager reconciles a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster.

Reviewer 2: ```suggestion
The Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. It also aggregates all module states into one easy-to-monitor resource, the `Kyma CR`.

This is done by 

1. generating the required Kyma module custom resources through a Manifest, which starts the Kyma module processing in Module Manager. 
2. reconciling a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster.

It should be noted that Lifecycle Manager is also responsible to propagate updates from a Module into the respective Manifest through the `ModuleTemplate`. This is a Custom Resource that is used by Lifecycle Manager as a scaffolding when initializing new Modules. It is also used as a base Output of the Module Bundling Process inside the CLI.

Whenever a new Module needs to be installable / registered in the Ecosystem, one only has to apply a new ModuleTemplate with the necessary information for generating the Module. Then it will be available for consumption in every Runtime Instance registered in the Control-Plane.

For every new module, Lifecycle Manager tracks the available ModuleTemplates in different release configurations and makes them available in a Catalog, which allows a browsable interface for all available modules. This Catalog is also synchronized into all Runtime Clusters so users can enable Modules as Runtime Administrators without needing Access to the Control Plane.

Whenever Modules are Updated and a new Release is triggered, Lifecycle Manager will make sure to propagate the correct upgrade information to the Manifest, which in turn leads to updates in the Runtime Cluster.
```

Author: Are the lines numbered with `1.` and `2.` subsequent steps, alternatives to each other (unlikely), or concurring processes?

```suggestion
Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. It also aggregates all module states into one easy-to-monitor resource, the `Kyma CR`, with the following strategies:

1. Generating the required Kyma module custom resources through a manifest, which starts the Kyma module processing in Module Manager. 

2. Reconciling a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster.

Lifecycle Manager is also responsible for propagating updates from a module into the respective manifest through the ModuleTemplate. Whenever modules are updated and a new release is triggered, Lifecycle Manager propagates the correct upgrade information to the manifest, which in turn leads to updates in the runtime cluster.

## 5. Module Template

The ModuleTemplate is a custom resource that is used by Lifecycle Manager as a scaffolding when initializing new Modules. It is also used as a base output of the module bundling process inside the CLI.

Whenever a new module must be installable or registered in the ecosystem, users only have to apply a new ModuleTemplate with the necessary information for generating the module. Then, it is available for consumption in every runtime instance registered in the control plane.

For every new module, Lifecycle Manager tracks the available ModuleTemplates in different release configurations and makes them available in a catalog, so users have a browsable interface for all available modules. This catalog is also synchronized into all runtime clusters so users can enable modules as runtime administrators without needing access to the control plane.
```

---
title: Modularization in Kyma
---

> With Kyma’s modular approach, you can install exactly the Kyma components you want, making the installation very lightweight and adjusted to your business needs. Kyma uses three system components to support the modular approach: The Module Manager, the Lifecycle Manager, and the Runtime Watcher.

Reviewer 3: ```suggestion
With Kyma’s modular approach, you can install exactly the Kyma components you want, making the installation very lightweight and adjusted to your business needs. Kyma uses three system components to support the modular approach: Module Manager, Lifecycle Manager, and Runtime Watcher.
```

## Kyma’s Modular Approach
<!-- diagram still WIP! -->
![](assets/modular_approach.svg)
<!-- alternative draft: -->
![](assets/modular_approach2.svg)

## 1. Kyma Module

A Kyma module contains all information required to install and run the associated components in a Kyma runtime:

* One or more renderable Kubernetes manifests of the module, based on Kustomize and/or Helm
* A default configuration for module chart resource installation
* Optionally, further layers that can enhance, or be required by, the operator of a module

All these assets are bundled into a single container image using the [OCI image specification](https://github.com/opencontainers/image-spec).

> > **TIP:** Learn how to build a module(ADD_LINK).

Reviewer 3: Can you add the link, or is it not available for now?

Author: I don't think it exists already :/

Reviewer 4: Currently, it'S this: https://github.com/kyma-project/lifecycle-manager/blob/main/docs/developer/creating-test-environment.md#create-module-operator-and-bundle-module

BUT, there is a bigger change happening which will simplify things massively (see [issue](https://github.com/kyma-project/lifecycle-manager/issues/226))

Author: So we don't link it and wait for the simplified version?

Reviewer 2: I suggest you link https://github.com/kyma-project/lifecycle-manager/tree/main/samples/template-operator as its our official reference implementation with step-by-step guides for setting up the operator and building the module. We will change this as well once the CLI is adapted but for this you dont have to adjust the link. There will be a move of this repository in a few weeks probably out of lifecycle-manager for easier access, then we will need to revisit this article

Author: ```suggestion
> **TIP:** Learn how to [build a module](https://github.com/kyma-project/lifecycle-manager/tree/main/samples/template-operator).
```

## 4. Lifecycle Manager

Lifecycle Manager is responsible for orchestrating Kyma module operators to process their respective resources on the Kyma runtime. It also aggregates all module states into one easy-to-monitor resource, the `Kyma CR`, with the following strategies:

1. Generating the required Kyma module custom resources through a manifest, which starts the Kyma module processing in Module Manager.

2. Reconciling a Kyma custom resource for each Kyma runtime, indicating the consolidated status of all modules configured for that Kyma cluster.

Lifecycle Manager is also responsible for propagating updates from a module into the respective manifest through the ModuleTemplate. Whenever modules are updated and a new release is triggered, Lifecycle Manager propagates the correct upgrade information to the manifest, which in turn leads to updates in the runtime cluster.

## 5. ModuleTemplate

The ModuleTemplate is a custom resource that is used by Lifecycle Manager as a scaffolding when initializing new modules. It is also used as a base output of the module bundling process inside the CLI.

> Whenever a new module must be installable or registered in the ecosystem, users only have to apply a new ModuleTemplate with the necessary information for generating the module. Then, it is available for consumption in every runtime instance registered in the control plane.

Reviewer 3: ```suggestion
Whenever a new module must be installed or registered in the ecosystem, you only have to apply a new ModuleTemplate with the necessary information to generate the module. Then, it is available for consumption in every runtime instance registered in the control plane.
```

Author: I'm not quite sure whether it must be installed, or whether it just must be possible to install it.
@tobiscr @jakobmoellersap @adityabhatia can you reply?

Reviewer 2: possible to install. ModuleTemplate only enables installation, it doesnt actually perform the installation

Reviewer 3: Got it. I updated the comment.

> For every new module, Lifecycle Manager tracks the available ModuleTemplates in different release configurations and makes them available in a catalog, which users can browse to see all available modules. This catalog is also synchronized into all runtime clusters, so users can enable modules as runtime administrators without needing access to the control plane.

Reviewer 3: ```suggestion
For every new module, Lifecycle Manager tracks the available ModuleTemplates in different release configurations and makes them available in a catalog, which you can browse to see all available modules. This catalog is also synchronized into all runtime clusters, so you can enable modules as runtime administrators without needing access to the control plane.
```
